diffrax/_solver/srk.py

```diff
@@ -0,0 +1,740 @@
+    SRK for solving SDEs with any type of noise (i.e. non-additive).
+    """
+
+    a: tuple[np.ndarray, ...]
+    b: Float[np.ndarray, " s"]
+
+    def check(self):
+        assert self.b.ndim == 1
+        assert all((i + 1,) == a_i.shape for i, a_i in enumerate(self.a))
+        assert self.b.shape[0] == len(self.a) + 1
+        return self.b.shape[0]
+
+
+class GeneralCoeffsWithError(_AbstractGeneralCoeffs):
+    """General coefficients for either the Brownian increment or its Lévy areas in an
+    SRK for solving SDEs with any type of noise (i.e. non-additive). Use this subclass
+    when the SRK provides an embedded method for error estimation.
+    """
+
+    a: tuple[np.ndarray, ...]
+    b: Float[np.ndarray, " s"]
+    b_error: Float[np.ndarray, " s"]
+
+    def check(self):
+        assert self.b.ndim == 1
+        assert all((i + 1,) == a_i.shape for i, a_i in enumerate(self.a))
+        assert self.b.shape[0] == len(self.a) + 1
+        assert self.b_error.ndim == 1
+        assert self.b_error.shape == self.b.shape
+        return self.b.shape[0]
+
+
+COEFFS = TypeVar("COEFFS", bound=AbstractStochasticCoeffs)
+
+
+class AbstractStochasticTableau(eqx.Module, Generic[COEFFS]):
+    r"""Part of a `StochasticButcherTableau` that represents the coefficients
+    for the Brownian increment and possibly its Levy areas."""
+
+    coeffs_w: eqx.AbstractVar[COEFFS]
+
+    @property
+    def a_w(self):
+        """The coefficients in front of the Brownian increment at each stage."""
+        return self.coeffs_w.a
+
+    @property
+    def b_w(self):
+        """The coefficient for the Brownian increment when computing the output."""
+        return self.coeffs_w.b
+
+    @property
+    def get_b_error_w(self) -> Optional[Float[np.ndarray, " s"]]:
+        if isinstance(self.coeffs_w, GeneralCoeffsWithError):
+            return self.coeffs_w.b_error
+        else:
+            return None
+
+    @property
+    def additive_noise(self):
+        return isinstance(self.coeffs_w, AdditiveCoeffs)
+
+    @property
+    def has_error_estimate(self):
+        return isinstance(self.coeffs_w, GeneralCoeffsWithError)
+
+    def check(self):
+        raise NotImplementedError
+
+
+class _AbstractSTLATableau(AbstractStochasticTableau[COEFFS]):
+    coeffs_hh: eqx.AbstractVar[COEFFS]
+
+    @property
+    def a_hh(self):
+        """The coefficients in front of the space-time Lévy area at each stage."""
+        return self.coeffs_hh.a
+
+    @property
+    def b_hh(self):
+        """The coefficient for the space-time Lévy area when computing the output."""
+        return self.coeffs_hh.b
+
+    @property
+    def get_b_error_hh(self) -> Optional[Float[np.ndarray, " s"]]:
+        if isinstance(self.coeffs_hh, GeneralCoeffsWithError):
+            return self.coeffs_hh.b_error
+        else:
+            return None
+
+
+class _AbstractSTTLATableau(_AbstractSTLATableau[COEFFS]):
```

Comment:
+1 for these abstract classes btw, I can see how you're using the pattern we discussed earlier!

Reply:
Thanks! To be honest it was quite a journey haha. A very illuminating one though.